src/repr/type.ml

```diff
@@ -60,6 +39,31 @@ let triple a b c = Tuple (Triple (a, b, c))
 let option a = Option a
 let boxed t = Boxed t
 
+let pre_hash t =
+  let rec aux : type a. a t -> a encode_bin = function
+    | Self s -> aux s.self_fix
+    | Map m ->
+        let dst = unstage (aux m.x) in
+        stage (fun v -> dst (m.g v))
+    | Custom c -> c.pre_hash
+    | t -> Type_binary.Unboxed.encode_bin t
+  in
+  aux t
+
+let short_hash = function
```

Comment:
I don't think you need to move these functions (`pre_hash` and `short_hash`); its harder to review the difference like this.